### PR TITLE
storage: prevent tag overwrites controlled via configuration

### PR DESCRIPTION
**Pull request by laurentiuNiculae**
Signed-off-by: Laurentiu Niculae niculae.laurentiu1@gmail.com
**What type of PR is this?**

**Which issue does this PR fix**:
[Issue #283](https://github.com/project-zot/zot/issues/283)

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:

**Does this change require updates to the CNI daemonset config files to work?**:

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/storage/storage_fs.go [hunk 1]
```diff
@@ -638,6 +640,11 @@ func (is *ImageStoreFS) PutImageManifest(repo string, reference string, mediaTyp
 				break
 			}
 			// manifest contents have changed for the same tag,
+
+			if !is.allowOverwrite {
```
Comment (andaaron) [inline pkg/storage/storage_fs.go:644]: We need unit tests to cover this logic.

File: pkg/storage/storage_fs.go [hunk 2]
```diff
@@ -55,16 +55,17 @@ type StoreController struct {
...
-	log         zerolog.Logger
-	metrics     monitoring.MetricServer
+	rootDir        string
+	lock           *sync.RWMutex
+	blobUploads    map[string]BlobUpload
+	cache          *Cache
+	gc             bool
+	dedupe         bool
+	commit         bool
+	allowOverwrite bool
```
Comment (andaaron) [inline pkg/storage/storage_fs.go:65]: Don't we need a similar change in zot/pkg/storage/s3/storage.go?

File: pkg/api/config/config.go [hunk 3]
```diff
@@ -68,6 +68,7 @@ type HTTPConfig struct {
 	Auth             *AuthConfig
 	RawAccessControl map[string]interface{} `mapstructure:"accessControl,omitempty"`
 	Realm            string
+	AllowOverwrites  bool             `mapstructure:",omitempty"`
```
Comment (andaaron) [inline pkg/api/config/config.go:71]: 1. @ we though this new configuration should be under `HTTPConfig`. Is this ok, or do you think it should be under `GlobalStorageConfig`?
2. Please add an example of such a configuration under `zot/examples`

File: pkg/api/routes.go [hunk 4]
```diff
@@ -451,6 +451,9 @@ func (rh *RouteHandler) UpdateManifest(response http.ResponseWriter, request *ht
 		} else if errors.Is(err, zerr.ErrBlobNotFound) {
 			WriteJSON(response, http.StatusBadRequest,
 				NewErrorList(NewError(BLOB_UNKNOWN, map[string]string{"blob": digest})))
+		} else if errors.Is(err, zerr.ErrCannotOverwrite) {
+			WriteJSON(response, http.StatusMethodNotAllowed,
+				NewErrorList(NewError(DENIED, map[string]string{})))
```
[the remark this answers is not in the document]
Comment (rchincha) [reply pkg/api/routes.go:456]: Yes, sounds reasonable

**Comment by rchincha** [comment]
Also consider adding this flag to the "access control" config section which allows for per-repo settings.

**Comment by rchincha** [comment]
@laurentiuNiculae pls sign this PR

File: pkg/cli/root.go [hunk 5]
```diff
@@ -412,3 +414,28 @@ func LoadConfiguration(config *config.Config, configPath string) error {
...
 	return nil
 }
+
+func isDefaultPolicyConfig(cfg *config.Config) bool {
+	adminPolicy := cfg.AccessControl.AdminPolicy
+
+	log.Info().Msg("checking if default authorization is possible")
+
+	if len(adminPolicy.Actions)+len(adminPolicy.Users) > 0 {
+		log.Info().Msg("default policy config can't have admin policy")
```
Comment (peusebiu) [inline pkg/cli/root.go:424]: I think this logs are misleading, same with below.

Maybe add a log.Info().Msg("authz with only default policies detected")

File: pkg/cli/root.go [hunk 6]
```diff
@@ -214,10 +214,12 @@ func validateConfiguration(config *config.Config) error {
...
 	if config.HTTP.RawAccessControl != nil {
 		if config.HTTP.Auth == nil || (config.HTTP.Auth.HTPasswd.Path == "" && config.HTTP.Auth.LDAP == nil) {
-			log.Error().Err(errors.ErrBadConfig).
-				Msg("access control config requires httpasswd or ldap authentication to be enabled")
-
-			return errors.ErrBadConfig
+			// checking for default authorization config: no users, no policies but default policy
+			if !isDefaultPolicyConfig(config) {
+				log.Error().Err(errors.ErrBadConfig).
+					Msg("access control config requires httpasswd or ldap authentication to be enabled")
```
Comment (peusebiu) [inline pkg/cli/root.go:220]: or only default policies to be used

File: pkg/api/controller_test.go [hunk 7]
```diff
@@ -2285,6 +2363,200 @@ func TestAuthorizationWithBasicAuth(t *testing.T) {
 	})
 }
 
+func TestAuthorizationWithDefaultPolicy(t *testing.T) {
```
Comment (peusebiu) [inline pkg/api/controller_test.go:2366]: withOnly

File: pkg/api/config/config.go [hunk 8]
```diff
@@ -140,7 +140,7 @@ func New() *Config {
 		Commit:          Commit,
 		BinaryType:      BinaryType,
 		Storage:         GlobalStorageConfig{GC: true, GCDelay: storage.DefaultGCDelay, Dedupe: true},
-		HTTP:            HTTPConfig{Address: "127.0.0.1", Port: "8080"},
+		HTTP:            HTTPConfig{Address: "127.0.0.1", Port: "8080", Auth: &AuthConfig{FailDelay: 0}},
```
Comment (peusebiu) [inline pkg/api/config/config.go:143]: why do you need this change?

File: pkg/cli/root.go (hunk 5 above)
Comment (laurentiuNiculae) [reply pkg/cli/root.go]: `func isDefaultPolicyConfig(cfg *config.Config) bool {
	adminPolicy := cfg.AccessControl.AdminPolicy

	log.Info().Msg("checking if default authorization is possible")

	if len(adminPolicy.Actions)+len(adminPolicy.Users) > 0 {
		log.Info().Msg("admin policy detected, default authorization disabled")

		return false
	}

	for _, repository := range cfg.AccessControl.Repositories {
		for _, policy := range repository.Policies {
			if len(policy.Actions)+len(policy.Users) > 0 {
				log.Info().Interface("repository", repository).
					Msg("repository with non-empty policy detected, default authorization disabled")

				return false
			}
		}
	}

	log.Info().Msg("default authorization detected")

	return true
}`
Are these logs more intuitive?

File: pkg/api/config/config.go (hunk 8 above)
Comment (peusebiu) [reply pkg/api/config/config.go:143]: I see, ok

File: pkg/cli/root.go [hunk 9]
```diff
@@ -214,10 +214,13 @@ func validateConfiguration(config *config.Config) error {
...
 		if config.HTTP.Auth == nil || (config.HTTP.Auth.HTPasswd.Path == "" && config.HTTP.Auth.LDAP == nil) {
-			log.Error().Err(errors.ErrBadConfig).
-				Msg("access control config requires httpasswd or ldap authentication to be enabled")
-
-			return errors.ErrBadConfig
+			// checking for default authorization config: no users, no policies but default policy
+			if !isDefaultPolicyConfig(config) {
+				log.Error().Err(errors.ErrBadConfig).
+					Msg("access control config requires httpasswd, ldap authentication or" +
+						"only default policies to be used to be enabled")
```
Comment (peusebiu) [inline pkg/cli/root.go:221]: access control config requires htpasswd or ldap authentication or using only 'defaultPolicy' policies. :D

**Review by peusebiu — DISMISSED** [review]

File: pkg/api/authz.go [hunk 10]
```diff
@@ -165,7 +165,7 @@ func matchesRepo(globPatterns map[string]bool, repository string) bool {
 	return allowed
 }
 
-func AuthzHandler(ctlr *Controller) mux.MiddlewareFunc {
+func AuthzHandler(ctlr *Controller, defaultAuthz bool) mux.MiddlewareFunc {
```
Comment (rchincha) [inline pkg/api/authz.go:168]: Add a comment for what "defaultAuthz" is for?
There is a authZ policy but no authN config.

Comment (rchincha) [reply pkg/api/authz.go:168]: Would add a comment?

**Review by rchincha — APPROVED** [review]
lgtm

**Review by peusebiu — APPROVED** [review]
LGTM

**Pull request merged**